### DISCUSSION TITLE
Disconnect event when underlying protocol state is CLOSED

### DISCUSSION THREAD

**Discussion by kosciej16** [miguelgrinberg/Flask-SocketIO#1633] [Q&A]
I have problem with understanding what happens behind scenes when connection is CLOSED (e.g. due to network problems).

I run simple server and connect with client, I am getting Ping/Pong normally:
```
[INFO] engineio.server: pYeMelv-SFQvfxuCAAAE: Sending packet PING data None
[DEBUG] websockets.protocol: server > Frame(fin=True, opcode=1, data=b'2', rsv1=False, rsv2=False, rsv3=False)
[DEBUG] websockets.protocol: server - event = data_received(<7 bytes>)
[DEBUG] websockets.protocol: server < Frame(fin=True, opcode=1, data=b'3', rsv1=False, rsv2=False, rsv3=False)
[INFO] engineio.server: pYeMelv-SFQvfxuCAAAE: Received packet PONG data
```
Now I cut client connection (using ctrl-c to stop the process in terminal, but I guess no internet connection do the same). I would expect **at some point disconnect will be called**, but apparently that's not the case:
```
[DEBUG] websockets.protocol: server - event = data_received(<15 bytes>)
[DEBUG] websockets.protocol: server < Frame(fin=True, opcode=8, data=b'\x03\xe8', rsv1=False, rsv2=False, rsv3=False)
[DEBUG] websockets.protocol: server - state = CLOSING
[DEBUG] websockets.protocol: server > Frame(fin=True, opcode=8, data=b'\x03\xe8', rsv1=False, rsv2=False, rsv3=False)
[DEBUG] websockets.protocol: server x half-closing TCP connection
[DEBUG] websockets.protocol: server - event = data_received(<8 bytes>)
[DEBUG] websockets.protocol: server - event = eof_received()
[DEBUG] websockets.protocol: server - event = connection_lost(None)
[DEBUG] websockets.protocol: server - state = CLOSED
[DEBUG] websockets.protocol: server x code = 1000, reason = [no reason]
```

1) How to add callback to call **disconnect** on that scenario?
2) WIll it disconnect client while wifi-problems too?

Thanks for help!

**Comment by miguelgrinberg** [comment]
You are asking this question generically, but it cannot be answered that way. This package supports a lot of different websocket implementations and the differences in their implementations can have an effect on how this package behaves. You also don't show what code you are using to test, so those logs do not have context that I need to be able to interpret them correctly.

So, please provide an update with the following:

- Server and client code you are using to test
- Output of `pip freeze`

**Reply by kosciej16** [reply]
@miguelgrinberg Thanks a lot for your reply
When I tried to create minimal example of my problem I couldn't reproduce it and indeed disconnect event is sent in the moment I would expect it. Not sure what was wrong earlier and sorry for taking your time, thank again!
